src/Application.php

```diff
@@ -73,23 +75,57 @@ class Application extends BaseApplication
      */
     protected $adapterFactory;
 
-    public function __construct(AdapterFactory $adapterFactory, $name = 'Gush', $version = '@package_version@')
+    public function __construct(AdapterFactory $adapterFactory, Config $config, $version = '@package_version@')
     {
         if ('@'.'package_version@' !== $version) {
             $version = ltrim($version, 'v');
         }
 
-        try {
-            // try setting the config as early as possible, so all the subscribers and helpers can use it
-            $this->config = Factory::createConfig();
-        } catch (\RuntimeException $exception) {
-            // if Gush is not yet configured, then just catch the exception and move along
-        }
+        parent::__construct('Gush', $version);
+
+        $this->adapterFactory = $adapterFactory;
+        $this->config = $config;
+
+        $helperSet = $this->getHelperSet();
+
+        // the parent dispatcher is private and has
+        // no accessor, so we set it here so we can access it.
+        $this->dispatcher = new EventDispatcher();
 
-        $helperSet = $this->getDefaultHelperSet();
+        $this->dispatcher->addSubscriber(new TableSubscriber());
+        $this->dispatcher->addSubscriber(
+            new GitRepoSubscriber(
+                $this,
+                $helperSet->get('git'),
+                $helperSet->get('git_config'),
+                $helperSet->get('gush_style')
+            )
+        );
+
+        $this->dispatcher->addSubscriber(
```

Comment:
All the subscribers can be moved to a `getEventSubscribers` (or something) method, to clean up the constructor a bit.

Then also maybe we can add a option in the `gush.yml` file to add custom event listeners:

```yaml
listeners:
    - ./my/custom/directory/Listerner.php
```
This will allow you to extend gush with some custom logic